lang/java/protobuf/src/test/java/org/apache/avro/protobuf/TestProtoConversions.java

```diff
@@ -0,0 +1,159 @@
+
+    long instant = May_28_2015_21_46_53_221.getTimeInMillis();
+    Timestamp May_28_2015_21_46_53_221_ts = Timestamp.newBuilder().setSeconds(1432849613L).setNanos(221000000).build();
+
+    Timestamp tsFromInstant = conversion.fromLong(instant, TIMESTAMP_MILLIS_SCHEMA, LogicalTypes.timestampMillis());
+    long roundTrip = conversion.toLong(tsFromInstant, TIMESTAMP_MILLIS_SCHEMA, LogicalTypes.timestampMillis());
+    Assert.assertEquals("Round-trip conversion should work", instant, roundTrip);
+
+    Assert.assertEquals("Known timestamp should be correct", May_28_2015_21_46_53_221_ts,
+        conversion.fromLong(instant, TIMESTAMP_MILLIS_SCHEMA, LogicalTypes.timestampMillis()));
+    Assert.assertEquals("Known timestamp should be correct", instant,
+        (long) conversion.toLong(May_28_2015_21_46_53_221_ts, TIMESTAMP_MILLIS_SCHEMA, LogicalTypes.timestampMillis()));
+  }
+
+  @Test
+  public void testTimestampMicrosConversion() {
+    TimestampMicrosConversion conversion = new TimestampMicrosConversion();
+
+    long instant = May_28_2015_21_46_53_221.getTimeInMillis() * 1000 + 843;
+    Timestamp May_28_2015_21_46_53_221_843_ts = Timestamp.newBuilder().setSeconds(1432849613L).setNanos(221843000)
+        .build();
+
+    Timestamp tsFromInstant = conversion.fromLong(instant, TIMESTAMP_MICROS_SCHEMA, LogicalTypes.timestampMicros());
+    long roundTrip = conversion.toLong(tsFromInstant, TIMESTAMP_MICROS_SCHEMA, LogicalTypes.timestampMicros());
+    Assert.assertEquals("Round-trip conversion should work", instant, roundTrip);
+
+    Assert.assertEquals("Known timestamp should be correct", May_28_2015_21_46_53_221_843_ts,
+        conversion.fromLong(instant, TIMESTAMP_MICROS_SCHEMA, LogicalTypes.timestampMicros()));
+    Assert.assertEquals("Known timestamp should be correct", instant, (long) conversion
+        .toLong(May_28_2015_21_46_53_221_843_ts, TIMESTAMP_MICROS_SCHEMA, LogicalTypes.timestampMicros()));
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void testTimestampMillisConversionSecondsLowerLimit() throws Exception {
+    TimestampMillisConversion conversion = new TimestampMillisConversion();
+    long exceeded = (ProtoConversions.SECONDS_LOWERLIMIT - 1) * 1000;
+    conversion.fromLong(exceeded, TIMESTAMP_MILLIS_SCHEMA, LogicalTypes.timestampMillis());
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void testTimestampMillisConversionSecondsUpperLimit() throws Exception {
+    TimestampMillisConversion conversion = new TimestampMillisConversion();
+    long exceeded = (ProtoConversions.SECONDS_UPPERLIMIT + 1) * 1000;
+    conversion.fromLong(exceeded, TIMESTAMP_MILLIS_SCHEMA, LogicalTypes.timestampMillis());
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void testTimestampMillisConversionNanosLowerLimit() throws Exception {
+    TimestampMillisConversion conversion = new TimestampMillisConversion();
+    long exceeded = -1L;
```

Comment:
This should actually work, because -0.001 seconds should be split into -1 whole seconds and +0.999 split seconds, which conforms to the rule that the split seconds part must be >= 0 and < 1. It is also a completely valid timestamp within the allowed range (1969-12-31 23:59:59.999).

Reply:
This sounds like a good point, thanks @zivanfi ! I'm afraid that we can't handle fraction of seconds before Unix epoch - I think this should be addressed.